docassemble/ALToolbox/al_income.py

```diff
@@ -0,0 +1,1015 @@
+            want to calcualte. E.g, to express a weekly period, use 52. Default is 1.
+        """
+        if times_per_year == 0:
+            return Decimal(0)
+
+        # If an item has its own period, use that instead
+        frequency_to_use = self.times_paid_per_year
+        if hasattr(item, "times_paid_per_year") and item.times_paid_per_year:
+            frequency_to_use = item.times_paid_per_year
+
+        is_hourly = False
+        # Override if item should be calculated hourly (like wages). Both the job
+        # and the item itself need to declare they should be calculated hourly
+        if self.is_hourly and hasattr(item, "is_hourly"):
+            is_hourly = item.is_hourly
+
+        # Conform to behavior of docassemble PeriodicValue
+        value = Decimal(0)
+        if hasattr(item, "value"):
+            value = Decimal(item.value)
+
+        # Use the appropriate cacluation
+        if is_hourly:
+            return (
+                value * Decimal(self.hours_per_period) * Decimal(frequency_to_use)
+            ) / Decimal(times_per_year)
+        else:
+            return (value * Decimal(frequency_to_use)) / Decimal(times_per_year)
+
+    def total(self, times_per_year=1, source=None):
+        """
+        Alias for ALItemizedJob.gross_total to integrate with ALIncomeList math.
+        """
+        return self.gross_total(times_per_year=times_per_year, source=source)
+
+    def gross_total(self, times_per_year=1, source=None):
+        """
+        Returns the sum of positive values (payments) for a given times_per_year.
+        You can filter the items by `source`. `source` can be a string or a list.
+        If you use sources from deductions, they will be ignored.
+
+        Args:
+        kwarg: times_per_year {str | num} (Optional) Number of times per year you
+            want to calcualte. E.g, to express a weekly period, use 52. Default is 1.
+        kwarg: source {str | [str]} (Optional) Source or list of sources of desired
+            item(s).
+        """
+        # self.to_add._trigger_gather()
+        total = Decimal(0)
+        if times_per_year == 0:
+            return total
+        # Make sure we're always working with a list of sources (names?)
+        sources = self.source_to_list(source=source)
+        # Add up all money coming in from a source
+        for key, value in self.to_add.elements.items():
+            if key in sources:
+                total += self.item_value_per_times_paid_per_year(
+                    value, times_per_year=times_per_year
+                )
+        return total
+
+    def deduction_total(self, times_per_year=1, source=None):
+        """
+        Returns the sum of money going out divided by a pay times_per_year as a
+        postive value. You can filter the items by `source`. `source` can be a
+        string or a list. If you use sources from money coming in, they will be
+        ignored.
```

Comment:
```suggestion
        Returns the total amount of "outgoing" values (deductions on a paycheck) for the specified time period.
        The total will be a positive value.
                
        You can filter the items by `source`. `source` can be a
        string or a list. If you use sources from money coming in, they will be
        ignored.
```

Reply:
I'm not sure what the "sources from money coming in" ignored refers to. If the source doesn't apply it's not used, right? Is it adding different context?

Reply:
It's just a warning that there's already a filter on this list that excludes money coming in (payments, etc.)